dali/util/shared_mem.cc

```diff
@@ -0,0 +1,188 @@
+  }
+  memcpy(temp_path, shm_path, kShmPathSize);
+  memcpy(temp_path + kShmPathSize, temp_filename_template, sizeof(temp_filename_template));
+  int temp_fd = mkstemp(temp_path);
+  if (temp_fd < 0) {
+    throw std::runtime_error("temporary file creation failed");
+  }
+  if (unlink(temp_path) != 0) {
+    throw std::runtime_error("couldn't unlink temporary file");
+  }
+  const char * temp_filename = temp_path + kShmPathSize;
+  fd_ = shm_open(temp_filename, O_CREAT | O_EXCL | O_RDWR, S_IRUSR | S_IWUSR);
+  if (fd_ < 0) {
+    throw std::runtime_error("shm_open call failed");
+  }
+  ::close(temp_fd);
+  // The memory chunks will be passed between processes by sending descriptors over the sockets,
+  // not through the filenames; this way in case of unexpected process exit we're not leaking any
+  // filelike objects in /dev/shm or /tmp. The OS guarantees to keep the memory as long as
+  // any process has fds or has mmaped the memory.
+  shm_unlink(temp_filename);
+}
+
+ShmFdWrapper::~ShmFdWrapper() {
+  close();
+}
+
+int ShmFdWrapper::get_fd() {
+  return fd_;
+}
+
+int ShmFdWrapper::close() {
+  int ret = 0;
+  if (fd_ >= 0) {
+    ret = ::close(fd_);
+    fd_ = -1;
+  }
+  return ret;
+}
+
+MapMemWrapper::MapMemWrapper(int fd, uint64_t size)
+    : size_{size},
+      ptr_{static_cast<SharedMem::b_type *>(
+          mmap(nullptr, size_, PROT_WRITE | PROT_READ, MAP_SHARED, fd, 0))} {
+  if (ptr_ == MAP_FAILED) {
+    throw std::runtime_error("mmap failed");
+  }
+}
+
+MapMemWrapper::~MapMemWrapper() {
+  unmap();
+}
+
+uint8_t *MapMemWrapper::get_raw_ptr() {
+  return ptr_;
+}
+
+void MapMemWrapper::resize(uint64_t new_size) {
+  if (!ptr_) {
+    throw std::runtime_error("cannot resize the memory mapping, because no memory is mapped");
+  }
+  ptr_ = static_cast<SharedMem::b_type *>(mremap(ptr_, size_, new_size, MREMAP_MAYMOVE));
+  if (ptr_ == MAP_FAILED) {
+    throw std::runtime_error("mremap failed");
+  }
+  size_ = new_size;
+}
+
+int MapMemWrapper::unmap() {
+  int ret = 0;
+  if (ptr_) {
+    ret = munmap(ptr_, size_);
+    ptr_ = nullptr;
+  }
+  return ret;
+}
+
+SharedMem::SharedMem(int fd, uint64_t size) : size_{size * sizeof(SharedMem::b_type)} {
+  if (fd >= 0) {
+    fd_ = std::make_unique<ShmFdWrapper>(fd);
+  } else {
+    fd_ = std::make_unique<ShmFdWrapper>();
+    if (ftruncate(fd_->get_fd(), size_) == -1) {
+      throw std::runtime_error("failed to resize shared memory");
+    }
+  }
+  mem_ = std::make_unique<MapMemWrapper>(fd_->get_fd(), size_);
+}
+
+uint64_t SharedMem::size() const {
+  return size_;
+}
+
+int SharedMem::fd() {
+  return !fd_ ? -1 : fd_->get_fd();
+}
+
+SharedMem::b_type *SharedMem::get_raw_ptr() {
+  return !mem_ ? nullptr : mem_->get_raw_ptr();
+}
+
+void SharedMem::resize(uint64_t size, bool trunc) {
+  size_ = size * sizeof(SharedMem::b_type);
+  if (trunc) {
+    if (ftruncate(fd_->get_fd(), size_) == -1) {
+      throw std::runtime_error("failed to resize shared memory");
+    }
+  }
```

Comment:
Does it provide any value?
```suggestion
  uint64_t new_size = size * sizeof(SharedMem::b_type);
  if (trunc && size_ > new_size) {
    if (ftruncate(fd_->get_fd(), size_) == -1) {
      throw std::runtime_error("failed to resize shared memory");
    }
  }
  size_ = new_size;
```

Reply:
Now I see `ftruncate` can extend as well. NVM